// portis/es/provider.js
import { isMobile, isLocalhost, randomId } from "./utils";
import { css } from './style';
var sdkVersion = '1.3.2';
var postMessages = {
    PT_RESPONSE: 'PT_RESPONSE',
    PT_HANDLE_REQUEST: 'PT_HANDLE_REQUEST',
    PT_GREEN_LIGHT: 'PT_GREEN_LIGHT',
    PT_SHOW_IFRAME: 'PT_SHOW_IFRAME',
    PT_HIDE_IFRAME: 'PT_HIDE_IFRAME',
    PT_USER_DENIED: 'PT_USER_DENIED',
    PT_USER_LOGGED_IN: 'PT_USER_LOGGED_IN',
    PT_PURCHASE_INITIATED: 'PT_PURCHASE_INITIATED',
    PT_ON_DATA: 'PT_ON_DATA',
    PT_SHOW_NOTIFICATION: 'PT_SHOW_NOTIFICATION',
    PT_HIDE_NOTIFICATION: 'PT_HIDE_NOTIFICATION',
};
var portisPayloadMethods = {
    SET_DEFAULT_EMAIL: 'SET_DEFAULT_EMAIL',
    SHOW_PORTIS: 'SHOW_PORTIS',
    CHANGE_NETWORK: 'CHANGE_NETWORK',
    SHOW_TX_DETAILS: 'SHOW_TX_DETAILS',
    UPDATE_WIDGET_LOCATION: 'UPDATE_WIDGET_LOCATION',
};
var PortisProvider = /** @class */ (function () {
    function PortisProvider(opts) {
        this.portisClient = 'https://app.portis.io';
        this.requests = {};
        this.queue = [];
        this.account = null;
        this.network = null;
        this.isPortis = true;
        this.events = [];
        if (!isLocalhost() && !opts.apiKey) {
            throw 'apiKey is missing. Please check your apiKey in the Portis dashboard: https://app.portis.io/dashboard';
        }
        if (opts.infuraApiKey && opts.providerNodeUrl) {
            throw 'Invalid parameters. \'infuraApiKey\' and \'providerNodeUrl\' cannot be both provided. Refer to the Portis documentation for more info.';
        }
        this.referrerAppOptions = {
            sdkVersion: sdkVersion,
            network: opts.network || 'mainnet',
            apiKey: opts.apiKey,
            infuraApiKey: opts.infuraApiKey,
            providerNodeUrl: opts.providerNodeUrl,
            scope: opts.scope,
        };
        this.elements = this.createIframe();
        this.listen();
    }
    PortisProvider.prototype.sendAsync = function (payload, cb) {
        this.enqueue(payload, cb);
    };
    PortisProvider.prototype.send = function (payload, cb) {
        if (cb) {
            this.sendAsync(payload, cb);
            return;
        }
        var result;
        switch (payload.method) {
            case 'eth_accounts':
                var account = this.account;
                result = account ? [account] : [];
                break;
            case 'eth_coinbase':
                result = this.account;
                break;
            case 'net_version':
                result = this.network;
                break;
            case 'eth_uninstallFilter':
                this.sendAsync(payload, function (_) { return _; });
                result = true;
                break;
            default:
                throw new Error("The Portis Web3 object does not support synchronous methods like " + payload.method + " without a callback parameter.");
        }
        return {
            id: payload.id,
            jsonrpc: payload.jsonrpc,
            result: result,
        };
    };
    PortisProvider.prototype.isConnected = function () {
        return true;
    };
    PortisProvider.prototype.setDefaultEmail = function (email, editable) {
        if (editable === void 0) { editable = true; }
        this.sendGenericPayload(portisPayloadMethods.SET_DEFAULT_EMAIL, [email, editable]);
    };
    PortisProvider.prototype.showPortis = function (callback) {
        this.sendGenericPayload(portisPayloadMethods.SHOW_PORTIS, [], callback);
    };
    PortisProvider.prototype.on = function (eventName, callback) {
        this.events.push({ eventName: eventName, callback: callback });
    };
    PortisProvider.prototype.enable = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.sendGenericPayload('eth_accounts', undefined, function (err, resp) {
                if (err) {
                    reject(err);
                }
                else {
                    resolve(resp.result);
                }
            });
        });
    };
    PortisProvider.prototype.changeNetwork = function (opts, callback) {
        if (opts.network) {
            this.referrerAppOptions.network = opts.network;
        }
        if (opts.infuraApiKey) {
            this.referrerAppOptions.infuraApiKey = opts.infuraApiKey;
        }
        if (opts.providerNodeUrl) {
            this.referrerAppOptions.providerNodeUrl = opts.providerNodeUrl;
        }
        this.sendGenericPayload(portisPayloadMethods.CHANGE_NETWORK, [opts.network, opts.infuraApiKey, opts.providerNodeUrl], callback);
    };
    PortisProvider.prototype.sendGenericPayload = function (method, params, callback) {
        if (params === void 0) { params = []; }
        if (callback === void 0) { callback = function (_) { return _; }; }
        var payload = {
            id: randomId(),
            jsonrpc: '2.0',
            method: method,
            params: params,
        };
        this.enqueue(payload, callback);
    };
    PortisProvider.prototype.createIframe = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var onload = function () {
                var mobile = isMobile();
                var wrapper = document.createElement('div');
                var iframe = document.createElement('iframe');
                var styleElem = document.createElement('style');
                var viewportMetaTag = document.createElement('meta');
                var notification = document.createElement('div');
                var notificationDetails = document.createElement('div');
                var notificationLogo = document.createElement('img');
                var notificationText = document.createElement('span');
                var showDetailsBtn = document.createElement('a');
                wrapper.className = mobile ? 'portis-mobile-wrapper' : 'portis-wrapper';
                iframe.className = mobile ? 'portis-mobile-iframe' : 'portis-iframe';
                notificationDetails.className = 'portis-notification-details';
                notification.className = mobile ? 'portis-mobile-notification' : 'portis-notification';
                notificationLogo.src = 'https://assets.portis.io/portis-logo/logo_64_64.png';
                notificationLogo.className = 'portis-notifiction-logo';
                iframe.src = _this.portisClient + "/send/?p=" + btoa(JSON.stringify(_this.referrerAppOptions));
                styleElem.innerHTML = css;
                viewportMetaTag.name = 'viewport';
                viewportMetaTag.content = 'width=device-width, initial-scale=1';
                showDetailsBtn.innerHTML = 'Details';
                showDetailsBtn.onclick = function () {
                    _this.sendGenericPayload(portisPayloadMethods.SHOW_TX_DETAILS);
                };
                showDetailsBtn.className = 'portis-notification-button';
                notificationDetails.appendChild(notificationLogo);
                notificationDetails.appendChild(notificationText);
                notification.appendChild(notificationDetails);
                notification.appendChild(showDetailsBtn);
                wrapper.appendChild(iframe);
                document.body.appendChild(wrapper);
                document.body.appendChild(notification);
                document.head.appendChild(styleElem);
                _this.portisViewportMetaTag = viewportMetaTag;
                _this.dappViewportMetaTag = _this.getDappViewportMetaTag();
                resolve({ wrapper: wrapper, iframe: iframe, notification: notification });
            };
            if (['loaded', 'interactive', 'complete'].indexOf(document.readyState) > -1) {
                onload();
            }
            else {
                window.addEventListener('load', onload.bind(_this), false);
            }
        });
    };
    PortisProvider.prototype.showIframe = function () {
        var _this = this;
        this.elements.then(function (elements) {
            elements.wrapper.style.display = 'block';
            _this.sendGenericPayload(portisPayloadMethods.UPDATE_WIDGET_LOCATION, [elements.wrapper.getBoundingClientRect()]);
            if (isMobile()) {
                document.body.style.overflow = 'hidden';
                _this.setPortisViewport();
            }
        });
    };
    PortisProvider.prototype.hideIframe = function () {
        var _this = this;
        this.elements.then(function (elements) {
            elements.wrapper.style.display = 'none';
            if (isMobile()) {
                document.body.style.overflow = 'inherit';
                _this.setDappViewport();
            }
        });
    };
    PortisProvider.prototype.showNotification = function (msg, showDetailsButton) {
        this.elements.then(function (elements) {
            var button = elements.notification.querySelector('a');
            if (button) {
                button.style.display = showDetailsButton ? 'initial' : 'none';
            }
            var span = elements.notification.querySelector('span');
            if (span) {
                span.innerText = msg;
                elements.notification.style.display = 'flex';
            }
        });
    };
    PortisProvider.prototype.hideNotification = function () {
        this.elements.then(function (elements) {
            var span = elements.notification.querySelector('span');
            if (span) {
                span.innerText = '';
                elements.notification.style.display = 'none';
            }
        });
    };
    PortisProvider.prototype.getDappViewportMetaTag = function () {
        var metaTags = document.head.querySelectorAll('meta[name=viewport]');
        return metaTags.length ? metaTags[metaTags.length - 1] : null;
    };
    PortisProvider.prototype.setPortisViewport = function () {
        document.head.appendChild(this.portisViewportMetaTag);
        if (this.dappViewportMetaTag) {
            this.dappViewportMetaTag.remove();
        }
    };
    PortisProvider.prototype.setDappViewport = function () {
        if (this.portisViewportMetaTag) {
            this.portisViewportMetaTag.remove();
        }
        if (this.dappViewportMetaTag) {
            document.head.appendChild(this.dappViewportMetaTag);
        }
    };
    PortisProvider.prototype.enqueue = function (payload, cb) {
        this.queue.push({ payload: payload, cb: cb });
        if (this.iframeReady) {
            this.dequeue();
        }
    };
    PortisProvider.prototype.dequeue = function () {
        if (this.queue.length == 0) {
            return;
        }
        var item = this.queue.shift();
        if (item) {
            var payload = item.payload;
            var cb = item.cb;
            var id = Array.isArray(payload) ?
                payload.map(function (i) { return i.id; }).join(':') :
                payload.id;
            this.sendPostMessage(postMessages.PT_HANDLE_REQUEST, payload);
            this.requests[id] = { payload: payload, cb: cb };
            this.dequeue();
        }
    };
    PortisProvider.prototype.sendPostMessage = function (msgType, payload) {
        this.elements.then(function (elements) {
            if (elements.iframe.contentWindow) {
                elements.iframe.contentWindow.postMessage({ msgType: msgType, payload: payload }, '*');
            }
        });
    };
    PortisProvider.prototype.listen = function () {
        var _this = this;
        window.addEventListener('message', function (evt) {
            if (evt.origin === _this.portisClient) {
                switch (evt.data.msgType) {
                    case postMessages.PT_GREEN_LIGHT: {
                        _this.iframeReady = true;
                        _this.dequeue();
                        break;
                    }
                    case postMessages.PT_RESPONSE: {
                        var id = Array.isArray(evt.data.response) ?
                            evt.data.response.map(function (i) { return i.id; }).join(':') :
                            evt.data.response.id;
                        _this.requests[id].cb(null, evt.data.response);
                        if (_this.requests[id].payload.method === 'eth_accounts' || _this.requests[id].payload.method === 'eth_coinbase') {
                            _this.account = evt.data.response.result[0];
                        }
                        if (_this.requests[id].payload.method === 'net_version') {
                            _this.network = evt.data.response.result;
                        }
                        _this.dequeue();
                        break;
                    }
                    case postMessages.PT_SHOW_IFRAME: {
                        _this.showIframe();
                        break;
                    }
                    case postMessages.PT_HIDE_IFRAME: {
                        _this.hideIframe();
                        break;
                    }
                    case postMessages.PT_SHOW_NOTIFICATION: {
                        _this.showNotification(evt.data.response.message, evt.data.response.showDetailsButton);
                        break;
                    }
                    case postMessages.PT_HIDE_NOTIFICATION: {
                        _this.hideNotification();
                        break;
                    }
                    case postMessages.PT_USER_DENIED: {
                        var id = evt.data.response ? evt.data.response.id : null;
                        if (id) {
                            _this.requests[id].cb(new Error('User denied transaction signature.'));
                        }
                        else {
                            _this.queue.forEach(function (item) { return item.cb(new Error('User denied transaction signature.')); });
                        }
                        _this.dequeue();
                        break;
                    }
                    case postMessages.PT_USER_LOGGED_IN: {
                        _this.events
                            .filter(function (event) { return event.eventName == 'login'; })
                            .forEach(function (event) { return event.callback({
                            provider: 'portis',
                            address: evt.data.response.address,
                            email: evt.data.response.email,
                        }); });
                        break;
                    }
                    case postMessages.PT_PURCHASE_INITIATED: {
                        _this.events
                            .filter(function (event) { return event.eventName == 'purchase-initiated'; })
                            .forEach(function (event) { return event.callback({
                            provider: 'portis',
                            purchaseId: evt.data.response.purchaseId,
                        }); });
                        break;
                    }
                    case postMessages.PT_ON_DATA: {
                        _this.events
                            .filter(function (event) { return event.eventName == 'data'; })
                            .forEach(function (event) { return event.callback(evt.data.response); });
                        break;
                    }
                }
            }
        }, false);
    };
    return PortisProvider;
}());
export { PortisProvider };
